ENH: Format `datetime.datetime` and `pd.Timestamp` objects in `pd.to_datetime`

aaossa opened the pull request:
- [X] Closes #49298
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [X] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

---

**Description of changes**

When passing `datetime.datetime` or `pd.Timestamp` objects to `pd.to_datetime`, while using the `format` argument, a `ValueError` is raised because those object does not match the expected format.

The implemented solution looks for these non-string objects in the `array_strptime` method, and then each object is processed to maintain its contents (which might differ from the expected `format` for strings), and then continuing processing as usual.

aaossa committed:
Parse `datetime` properly in `pd.to_datetime`

When applying `pd.to_datetime` on array-like structure that contain a
`datetime.datetime` object, while using the `format` argument, a
`ValueError` is raised because the `datetime.datetime` object does not
match the expected format.

The implemented solution looks for `datetime.datetime` instances in the
`array_strptime` method. If an instance of this type is found, it's
properly handled by the new `_parse_python_datetime_object`, which
returns the expected Numpy datetime object.

Signed-off-by: Antonio Ossa Guerra <aaossa@uc.cl>

aaossa committed:
Test `pd_to_datetime` in presence of `datetime`

Two tests are introduced to check that `datetime.datetime` objects are
skipped while parsing array-like structures to `pd.to_datetime` and
the `format` argument is being used.

Before the fix, all of these tests and their test cases failed, except
`test_to_datetime_arraylike_contains_pydatetime_utc` when cache was
`True` and the `init_constructor` was `Index`.

The test creates an input of data that must be formated and a
`datetime.datetime` object. Then, the values are passed to the
array-like `init_constructor` to be converted to datetime using the
`pd.to_datetime` method. The test must not fail, and the result must
match the expected data.

Signed-off-by: Antonio Ossa Guerra <aaossa@uc.cl>

aaossa committed:
Add whatsnew on `to_datetime` handling `datetime`

The introduced enhancement of handling `datetime.datetime` objects in
the input of `pd.to_datetime` when using the `format` argument is now
included in the whatsnew

Signed-off-by: Antonio Ossa Guerra <aaossa@uc.cl>

jbrockmendel commented on pandas/_libs/tslibs/strptime.pyx:
```diff
@@ -129,12 +130,22 @@ def array_strptime(ndarray[object] values, str fmt, bint exact=True, errors='rai
+        elif checknull_with_nat_and_na(val):
+            iresult[i] = NPY_NAT
+            continue
+        elif isinstance(val, datetime):
```
can you use PyDateTime_Check here

jbrockmendel replied:
probably need to cimport `import_datetime` and have a line that just says "import_datetime" (take a look at other files in this directory to see how that works)

aaossa replied:
Thanks, that worked! Unfortunately, changing the condition breaks support for `Timestamp` objects, breaking in the same way as originally reported for `datetime.datetime` objects. When using `isisntance(val, datetime)`, `Timestamp` objects are also considered. So, I'll use `isinstance` for now

MarcoGorelli replied:
which tests fail if you do that? I just tried applying 
```diff
diff --git a/pandas/_libs/tslibs/strptime.pyx b/pandas/_libs/tslibs/strptime.pyx
index 5ec8494f9f..4f935d51bb 100644
--- a/pandas/_libs/tslibs/strptime.pyx
+++ b/pandas/_libs/tslibs/strptime.pyx
@@ -5,6 +5,12 @@ from cpython.datetime cimport (
     datetime,
     tzinfo,
 )
+from cpython.datetime cimport (
+    PyDateTime_Check,
+    import_datetime,
+)
+
+import_datetime()
 
 from _thread import allocate_lock as _thread_allocate_lock
 
@@ -133,7 +139,7 @@ def array_strptime(ndarray[object] values, str fmt, bint exact=True, errors='rai
         elif checknull_with_nat_and_na(val):
             iresult[i] = NPY_NAT
             continue
-        elif isinstance(val, datetime):
+        elif PyDateTime_Check(val):
             val = val.strftime(fmt)
         else:
             val = str(val)
```

and then
```
pytest pandas/tests/tools/test_to_datetime.py
```
all passed

jbrockmendel replied:
id suggest taking a look at how we handle the `elif PyDateTime_Check(val)` case in `array_to_datetime`

aaossa replied:
You're right @MarcoGorelli , seems like I tested while applying multiple changes instead of isolating the usage of `PyDateTime_Check`. I'll update the PR

jbrockmendel commented on pandas/_libs/tslibs/strptime.pyx:
```diff
@@ -532,3 +543,29 @@ cdef tzinfo parse_timezone_directive(str z):
                      (microseconds // 60_000_000))
     total_minutes = -total_minutes if z.startswith("-") else total_minutes
     return pytz.FixedOffset(total_minutes)
+
```
one more newline

jbrockmendel commented on pandas/_libs/tslibs/strptime.pyx:
```diff
@@ -532,3 +543,29 @@ cdef tzinfo parse_timezone_directive(str z):
+    dts.min = dt.minute
+    dts.sec = dt.second
+    dts.us = dt.microsecond
+    dts.ps = 0  # Not enough precision in datetime objects (https://github.com/python/cpython/issues/59648)
```
what if we have a Timestamp object?

aaossa replied:
Fixed using `dts.ps = getattr(dt, "nanosecond", 0) * 1000`

aaossa replied:
This might need a test

mroeschke commented on pandas/_libs/tslibs/strptime.pyx:
```diff
@@ -129,12 +130,22 @@ def array_strptime(ndarray[object] values, str fmt, bint exact=True, errors='rai
+            iresult[i] = NPY_NAT
+            continue
+        elif isinstance(val, datetime):
+            iresult[i] = _parse_python_datetime_object(val, &dts)
```
Instead of this could we just do:

```
val = val.strftime(fmt)
```

And then the existing parsing code can be reused?

aaossa replied:
This does not work because `iresult` contains int64 date representations of the dates. IIRC, those values are used to represent the `Timestamp` objects internally

mroeschke replied:
Right but I mean here we convert the `datetime` to a `str` and the `for group_key in found_dict.iterkeys():` loop below will assign the `iresult[i]` value of the `datetime -> str`

aaossa replied:
I see it now. That works too. Seems like it depends on deciding if `datetime` and `Timestamp` objects should be left alone or formatted.

For example:

```python
>>> data = [Timestamp("2001-10-02 12:30:01"), "01/10/01"]
>>> 
>>> # Current approach: leave Timestamp/datetime objects alone
>>> pd.to_datetime(data, format="%d/%m/%y")
DatetimeIndex(['2001-10-02 12:30:01', '2001-10-01 00:00:00'], dtype='datetime64[ns]', freq=None)
>>> 
>>> # Using val = val.strftime(fmt): apply format on Timestamp/datetime objects
>>> pd.to_datetime(data, format="%d/%m/%y")
DatetimeIndex(['2001-10-02', '2001-10-01'], dtype='datetime64[ns]', freq=None)
```

I thought that leaving the objects alone (skipping them) was cleaner, but seems like it produces unexpected results. Applying `fmt` on `Timestamp` and `datetime` objects seems like a better approach now. Thanks for your suggestion!

aaossa committed:
Apply `fmt` on `Timestamp` and `datetime` objects

Instead of manually handling the `datetime.datetime` and `Timestamp`
objects by skipping formatting and leaving them untouched, apply the
passed format (`fmt`). This allow us to return a consistent output
while reusing existing parsing code

Signed-off-by: Antonio Ossa Guerra <aaossa@uc.cl>

aaossa committed:
Update tests to reflect new expected behavior

The new expecte bahavior is apply the format (`fmt`) on
`datetime.datetime` and `Timestamp` objects to create a consistent
output. This test includes some unexpected cases, like the limitation
of resolution on `datetime.datetime` objects (Python does not support
nanosecond resolution [1]) and also when formatting `Timestamp`
objects (same limitation [2])

[1]: https://github.com/python/cpython/issues/59648
[2]: https://github.com/pandas-dev/pandas/issues/29461

Signed-off-by: Antonio Ossa Guerra <aaossa@uc.cl>

aaossa committed:
Update whatsnew to describe new expected behavior

As described in the previous commit, the new expected behavior when
handling `datetime.datetime` and `Timestamp` objects changed, by
applying the passed format (`fmt`) instead of raising `ValueError`

Signed-off-by: Antonio Ossa Guerra <aaossa@uc.cl>

aaossa commented on pandas/tests/tools/test_to_datetime.py:
```diff
@@ -609,6 +609,119 @@ def test_to_datetime_dtarr(self, tz):
+        ]
+        result = to_datetime(init_constructor(case2), format="%m/%d/%y %H:%M:%S.%f")
+        expected_data = [
+            Timestamp("2001-10-01 12:00:01.123456"),
+            Timestamp("2001-10-02 12:30:01.123456"),
+            Timestamp("2001-10-03 13:00:01.123456789"),
```
This might seem weird but it has an explanation for each case:

* `Timestamp.strftime` does not support nanoseconds: https://github.com/pandas-dev/pandas/issues/29461
* `datetime.datetime` also has the same problem: https://github.com/python/cpython/issues/59648

MarcoGorelli commented on pandas/tests/tools/test_to_datetime.py:
```diff
@@ -609,6 +609,121 @@ def test_to_datetime_dtarr(self, tz):
+    )
+    def test_to_datetime_preserves_resolution_when_possible(self, data, expected):
+        # GH 49298
+        if not isinstance(data, str):
+            result = to_datetime([data])
+            tm.assert_equal(result, DatetimeIndex([data]))
```
why is the `if` statement necessary?

aaossa replied:
My bad, the condition is removed now (the test is still there)

MarcoGorelli commented on pandas/tests/tools/test_to_datetime.py:
```diff
@@ -609,6 +609,121 @@ def test_to_datetime_dtarr(self, tz):
+            (Series, Series),
+        ],
+    )
+    def test_to_datetime_arraylike_contains_pydatetime_and_timestamp(
```
not sure I follow, why is this (and the following) test necessary? I'd have thought that just some example which checks `to_datetime(arg, format=` where the input list contains both strings and `datetime.datetime` objects and `Timestamp` objects and `format` is specified.
Not sure I see why the constructor tests add, but I might be missing something

I'd also suggest adding a test where `format` is ISO8601 (as that goes down a different path), e.g. `%Y-%m-%d`

aaossa replied:
This test performs the comparison you mention: input with different types while verifying that the output matches the expected format. Constructors should not be necessary (at least in this test) so I'll remove them. Also, I'll add a case where `format` is SO8601 here.

The `test_to_datetime_arraylike_contains_pydatetime_and_timestamp_utc` test aimed to check that the timezone was preserved as expected. It can be removed, but I wanted to check consistency across different input types

MarcoGorelli reviewed (changes requested):
thanks for working on this

I've just left some comments

aaossa committed:
Use `PyDateTime_Check` instead of `isisntance`

The `PyDateTime_Check` function is used in other parts of the source
to properly check if an object is an instance of type
`PyDateTime_DateTimeType` or a subtype of the same type.

Signed-off-by: Antonio Ossa Guerra <aaossa@uc.cl>

aaossa committed:
Simplify tests by removing unnecesary parameters

Parametrized constructors add no information that is relevant in the
modified test, so they can be safely removed

Signed-off-by: Antonio Ossa Guerra <aaossa@uc.cl>

aaossa commented on pandas/tests/tools/test_to_datetime.py:
```diff
@@ -609,6 +609,88 @@ def test_to_datetime_dtarr(self, tz):
+        result = to_datetime([data], format="%m/%d/%y %H:%M:%S.%f")
+        tm.assert_equal(result, DatetimeIndex([expected]))
+
+    def test_to_datetime_arraylike_contains_pydatetime_and_timestamp(self):
```
This test still needs an ISO8601 case

aaossa replied:
Test added (case4)

MarcoGorelli commented on pandas/tests/tools/test_to_datetime.py:
```diff
@@ -609,6 +609,88 @@ def test_to_datetime_dtarr(self, tz):
+            ("10/01/03 12:00:01.123456789", Timestamp("2003-10-01 12:00:01.123456789")),
+        ],
+    )
+    def test_to_datetime_preserves_resolution_when_possible(self, data, expected):
```
I'm still confused as to why this test is necessary

Isn't the issue about `to_datetime` not skipping over objects that are already `dt.datetime`?

The subsequent test also mentions resolution, but I fail to see how it's related to the issue this PR is addressing

MarcoGorelli replied:
thanks for explaining, I'll take a closer look later in the week

MarcoGorelli replied:
I've had a look at that comment, but to be honest I don't see why skipping `Timestamp` or `datetime` objects would be inconsistent

I think this would be considered a bug:
```
In [24]: date_str
Out[24]: '10/01/2001 12:00:01.123456789'

In [25]: ts
Out[25]: Timestamp('2001-10-01 12:00:01.123456789')

In [26]: to_datetime([date_str, ts], format='%m/%d/%Y %H:%M:%S.%f')
Out[26]: DatetimeIndex(['2001-10-01 12:00:01.123456789', '2001-10-01 12:00:01.123456'], dtype='datetime64[ns]', freq=None)
```

I'd be in favour of:
- skipping objects that are already `Timestamp` or `datetime`
- erroring if there are `Timestamp` or `datetime` objects
but not of modifying them

aaossa replied:
The new changes skip objects that are already `Timestamp` or `datetime.datetime` objects, producing the expected output in this case: `DatetimeIndex(['2001-10-01 12:00:01.123456789', '2001-10-01 12:00:01.123456789'], dtype='datetime64[ns]', freq=None)`

aaossa committed:
Add ISO8601 test case

When using ISO8601 the format is inferred, so no format has to be
explicitly defined. This test case includes input strings in ISO8601
to assert the behavior when the input includes `Timestamp` and
`datetime.datetime` objects

Signed-off-by: Antonio Ossa Guerra <aaossa@uc.cl>

aaossa commented on doc/source/whatsnew/v2.0.0.rst:
```diff
@@ -49,6 +49,7 @@ Other enhancements
 - Fix ``test`` optional_extra by adding missing test package ``pytest-asyncio`` (:issue:`48361`)
 - :func:`DataFrame.astype` exception message thrown improved to include column name when type conversion is not possible. (:issue:`47571`)
 - :meth:`DataFrame.to_json` now supports a ``mode`` keyword with supported inputs 'w' and 'a'. Defaulting to 'w', 'a' can be used when lines=True and orient='records' to append record oriented json lines to an existing json file. (:issue:`35849`)
+- :func:`to_datetime` now handles ``datetime.datetime`` and :class:`Timestamp` and applies the ``format`` argument on them instead of raising a ``ValueError``. (:issue:`49298`)
```
I mean that the `strftime` method is used on each `datetime.datetime` and `Timestamp` object using the passed `format`. Maybe:

> `to_datetime` now formats `datetime.datetime`, `Timestamp` and `str` objects in the input to return a format-consistent output instead of raising a `ValueError`

aaossa committed:
Update whatsnew to explain expected result

Improve description by explaining the change of behavior in the result

Signed-off-by: Antonio Ossa Guerra <aaossa@uc.cl>

aaossa committed:
Do not modify `Timestamp` and `datetime` objects

To avoid introducing an unexpected behavior, we'll skip objects that
are already `pd.Timestamp` or `datetime.datetime`

Signed-off-by: Antonio Ossa Guerra <aaossa@uc.cl>

aaossa committed:
Update and simplify tests

Simplify the tests further by testing specific cases that are actually
relevant to the introduced behavior. The first test includes cases for
both custom and ISO8601, while the second tests timezone-aware objects
while forcing UTC and while leaving the offsets alone

Signed-off-by: Antonio Ossa Guerra <aaossa@uc.cl>

aaossa committed:
Update whatsnew describing new behavior

`pd.to_datetime` now simply skips `datetime.datetime` and `Timestamp`
objects when the `format` argument is present, instead of trying to
use the same format in the output

Signed-off-by: Antonio Ossa Guerra <aaossa@uc.cl>

MarcoGorelli commented on doc/source/whatsnew/v2.0.0.rst:
```diff
@@ -49,6 +49,7 @@ Other enhancements
 - Fix ``test`` optional_extra by adding missing test package ``pytest-asyncio`` (:issue:`48361`)
 - :func:`DataFrame.astype` exception message thrown improved to include column name when type conversion is not possible. (:issue:`47571`)
 - :meth:`DataFrame.to_json` now supports a ``mode`` keyword with supported inputs 'w' and 'a'. Defaulting to 'w', 'a' can be used when lines=True and orient='records' to append record oriented json lines to an existing json file. (:issue:`35849`)
+- :func:`to_datetime` now skips ``datetime.datetime`` and :class:`Timestamp` objects instead of raising a ``ValueError``. (:issue:`49298`)
```
let's note that it does this when the user passes `'format'` (I think it already did it in other cases?)

aaossa replied:
You're right, I'll work on it and include your example in a test

MarcoGorelli reviewed (changes requested):
Thanks for updating!

I think the handling of timezones might not be quite right:
```
In [2]: ts = Timestamp('1950-01-01').tz_localize('Europe/London')

In [3]: to_datetime([ts])
Out[3]: DatetimeIndex(['1950-01-01 00:00:00+00:00'], dtype='datetime64[ns, Europe/London]', freq=None)

In [4]: to_datetime([ts], format='%m-%d-%Y')
Out[4]: DatetimeIndex(['1950-01-01'], dtype='datetime64[ns]', freq=None)
```

aaossa committed:
Return localized object if a timezone is present

When using the `format` argument on `pd.to_timestamp`, and the input
contains localized `pd.Timestamp` or `datetime.datetime` objects, the
result should be localized too.

A bug was triggered when `format` did not contain `%z` or `%Z`,
because the `_array_strptime_with_fallback` function did not expect
any type different than string. Now this assumption is no longer
valid, since we now let `pd.Timestamp` and `datetime.datetime` objects
pass unmodified.

The fix is to change the condition that checks if returning a
localized output is necessary: instead of looking at the passed
`format`, return a localized result when any input object contains a
timezone.

Also, a new test is added to check this case. A case is included for a
`pd.Timestamp` and another for a `datetime.datetime` object, and each
compares the actual result with the expected result when the input is
localized and when is not.

Signed-off-by: Antonio Ossa Guerra <aaossa@uc.cl>

MarcoGorelli reviewed (changes requested):
Thanks for updating

It looks like now, mixing timezone-aware and timezone-naive is allowed if `format` is specified? Is this right? Shouldn't the behaviour be consistent with when `format` is not passed?
```python
In [1]: ts = Timestamp('1950-01-01').tz_localize('Europe/London')

In [2]: ts1 = Timestamp('1950-01-01')

In [3]: to_datetime([ts, ts1])
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In [3], line 1
----> 1 to_datetime([ts, ts1])

File ~/pandas-dev/pandas/core/tools/datetimes.py:1127, in to_datetime(arg, errors, dayfirst, yearfirst, utc, format, exact, unit, infer_datetime_format, origin, cache)
   1125         result = _convert_and_box_cache(argc, cache_array)
   1126     else:
-> 1127         result = convert_listlike(argc, format)
   1128 else:
   1129     result = convert_listlike(np.array([arg]), format)[0]

File ~/pandas-dev/pandas/core/tools/datetimes.py:440, in _convert_listlike_datetimes(arg, format, name, tz, unit, errors, infer_datetime_format, dayfirst, yearfirst, exact)
    438 assert format is None or infer_datetime_format
    439 utc = tz == "utc"
--> 440 result, tz_parsed = objects_to_datetime64ns(
    441     arg,
    442     dayfirst=dayfirst,
    443     yearfirst=yearfirst,
    444     utc=utc,
    445     errors=errors,
    446     require_iso8601=require_iso8601,
    447     allow_object=True,
    448 )
    450 if tz_parsed is not None:
    451     # We can take a shortcut since the datetime64 numpy array
    452     # is in UTC
    453     dta = DatetimeArray(result, dtype=tz_to_dtype(tz_parsed))

File ~/pandas-dev/pandas/core/arrays/datetimes.py:2204, in objects_to_datetime64ns(data, dayfirst, yearfirst, utc, errors, require_iso8601, allow_object, allow_mixed)
   2202 order: Literal["F", "C"] = "F" if flags.f_contiguous else "C"
   2203 try:
-> 2204     result, tz_parsed = tslib.array_to_datetime(
   2205         data.ravel("K"),
   2206         errors=errors,
   2207         utc=utc,
   2208         dayfirst=dayfirst,
   2209         yearfirst=yearfirst,
   2210         require_iso8601=require_iso8601,
   2211         allow_mixed=allow_mixed,
   2212     )
   2213     result = result.reshape(data.shape, order=order)
   2214 except OverflowError as err:
   2215     # Exception is raised when a part of date is greater than 32 bit signed int

File ~/pandas-dev/pandas/_libs/tslib.pyx:444, in pandas._libs.tslib.array_to_datetime()
    442 @cython.wraparound(False)
    443 @cython.boundscheck(False)
--> 444 cpdef array_to_datetime(
    445     ndarray[object] values,
    446     str errors='raise',

File ~/pandas-dev/pandas/_libs/tslib.pyx:553, in pandas._libs.tslib.array_to_datetime()
    551 found_naive = True
    552 if found_tz and not utc_convert:
--> 553     raise ValueError('Cannot mix tz-aware with '
    554                      'tz-naive values')
    555 if isinstance(val, _Timestamp):

ValueError: Cannot mix tz-aware with tz-naive values

In [4]: to_datetime([ts, ts1], format='%m-%d-%Y')
Out[4]: Index([1950-01-01 00:00:00+00:00, 1950-01-01 00:00:00], dtype='object')
```

aaossa committed:
Raise `ValueError` when tz-aware differs in input

Just like `to_datetime` when `format` is not passed, we should raise a
`ValueError` if the input contains some tz-aware objects and some
naive objects. The change is to identify if the `format` expects
tz-aware objects (check for `"%z"` and `"%Z"`) and then compare this
expectation with the `tzinfo` of each object. This modification
includes relevant tests.

Signed-off-by: Antonio Ossa Guerra <aaossa@uc.cl>

MarcoGorelli commented on pandas/_libs/tslibs/strptime.pyx:
```diff
@@ -122,19 +128,31 @@ def array_strptime(ndarray[object] values, str fmt, bint exact=True, errors='rai
-                val = str(val)
+                iresult[i] = pydatetime_to_dt64(val, &dts)
+                check_dts_bounds(&dts)
+            if val.tzinfo is None and expect_tz_aware:
+                raise ValueError("Cannot mix tz-aware with tz-naive values")
+            elif val.tzinfo is not None and not expect_tz_aware:
```
does it work to simplify this into
```suggestion
            if val.tzinfo is not None ^ expect_tz_aware:
```
?

MarcoGorelli reviewed (changes requested):
Thanks for sticking with this!

I'm still getting errors if we have timezone-aware Timestamps:
```python
In [9]: In [1]: ts = Timestamp('1950-01-01').tz_localize('Europe/London')
   ...: 
   ...: In [2]: ts1 = Timestamp('1950-01-01').tz_localize('Europe/London')

In [10]: to_datetime([ts, ts1], format='%d-%m-%Y')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In [10], line 1
----> 1 to_datetime([ts, ts1], format='%d-%m-%Y')

File ~/pandas-dev/pandas/core/tools/datetimes.py:1127, in to_datetime(arg, errors, dayfirst, yearfirst, utc, format, exact, unit, infer_datetime_format, origin, cache)
   1125         result = _convert_and_box_cache(argc, cache_array)
   1126     else:
-> 1127         result = convert_listlike(argc, format)
   1128 else:
   1129     result = convert_listlike(np.array([arg]), format)[0]

File ~/pandas-dev/pandas/core/tools/datetimes.py:432, in _convert_listlike_datetimes(arg, format, name, tz, unit, errors, infer_datetime_format, dayfirst, yearfirst, exact)
    429         format = None
    431 if format is not None:
--> 432     res = _to_datetime_with_format(
    433         arg, orig_arg, name, tz, format, exact, errors, infer_datetime_format
    434     )
    435     if res is not None:
    436         return res

File ~/pandas-dev/pandas/core/tools/datetimes.py:540, in _to_datetime_with_format(arg, orig_arg, name, tz, fmt, exact, errors, infer_datetime_format)
    537         return _box_as_indexlike(result, utc=utc, name=name)
    539 # fallback
--> 540 res = _array_strptime_with_fallback(
    541     arg, name, tz, fmt, exact, errors, infer_datetime_format
    542 )
    543 return res

File ~/pandas-dev/pandas/core/tools/datetimes.py:475, in _array_strptime_with_fallback(arg, name, tz, fmt, exact, errors, infer_datetime_format)
    472 utc = tz == "utc"
    474 try:
--> 475     result, timezones = array_strptime(arg, fmt, exact=exact, errors=errors)
    476 except OutOfBoundsDatetime:
    477     if errors == "raise":

File ~/pandas-dev/pandas/_libs/tslibs/strptime.pyx:151, in pandas._libs.tslibs.strptime.array_strptime()
    149     raise ValueError("Cannot mix tz-aware with tz-naive values")
    150 elif val.tzinfo is not None and not expect_tz_aware:
--> 151     raise ValueError("Cannot mix tz-aware with tz-naive values")
    152 result_timezone[i] = val.tzinfo
    153 continue

ValueError: Cannot mix tz-aware with tz-naive values
```

In this case, they're both `tz-aware`, so is it right that the error message refers to `tz-naive`?

aaossa commented:
> In this case, they're both `tz-aware`, so is it right that the error message refers to `tz-naive`?

To be clear, should this case raise a ValueError or not? Is the message the problem? What's the expected behavior in this case?

Relatedly, what should happen if there are no string to format? That argument would not have any effect, is a warning necessary/appropriate?

MarcoGorelli commented:
If they all have the same offset, then something like:
```python
In [3]: to_datetime(['01-01-2000 00:00:00 +01:00', '01-02-2000 00:00:00 +01:00'], format='%d-%m-%Y %H:%M:%S %z')
Out[3]: DatetimeIndex(['2000-01-01 00:00:00+01:00', '2000-02-01 00:00:00+01:00'], dtype='datetime64[ns, pytz.FixedOffset(60)]', freq=None)
```

If they have different offsets:
```
In [4]: to_datetime(['01-01-2000 00:00:00 +01:00', '01-02-2000 00:00:00 +02:00'], format='%d-%m-%Y %H:%M:%S %z')
Out[4]: Index([2000-01-01 00:00:00+01:00, 2000-02-01 00:00:00+02:00], dtype='object')
```

> Relatedly, what should happen if there are no string to format? That argument would not have any effect, is a warning necessary/appropriate?

🤔 that's a good one - tempted to say that we should just let the argument not have any effect, wouldn't bother with a warning

MarcoGorelli commented:
I'd suggest looking at

https://github.com/pandas-dev/pandas/blob/f3c46cd0899d5e11e0602798d9390c90e51e9ba7/pandas/_libs/tslib.pyx#L442

and perhaps making a function to factor out from there, because in the ISO8601 case then it seems that datetime / Timestamp objects are handled fine:

```python
In [10]: to_datetime(['2000-01-01', ts], format='%Y-%m-%d')
Out[10]: DatetimeIndex(['2000-01-01', '1950-01-01'], dtype='datetime64[ns]', freq=None)

In [11]: to_datetime(['2000-01-01', ts], format='%Y-%d-%m')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In [11], line 1
----> 1 to_datetime(['2000-01-01', ts], format='%Y-%d-%m')

File ~/tmp/.venv/lib/python3.10/site-packages/pandas/core/tools/datetimes.py:1100, in to_datetime(arg, errors, dayfirst, yearfirst, utc, format, exact, unit, infer_datetime_format, origin, cache)
   1098         result = _convert_and_box_cache(argc, cache_array)
   1099     else:
-> 1100         result = convert_listlike(argc, format)
   1101 else:
   1102     result = convert_listlike(np.array([arg]), format)[0]

File ~/tmp/.venv/lib/python3.10/site-packages/pandas/core/tools/datetimes.py:430, in _convert_listlike_datetimes(arg, format, name, tz, unit, errors, infer_datetime_format, dayfirst, yearfirst, exact)
    427         format = None
    429 if format is not None:
--> 430     res = _to_datetime_with_format(
    431         arg, orig_arg, name, tz, format, exact, errors, infer_datetime_format
    432     )
    433     if res is not None:
    434         return res

File ~/tmp/.venv/lib/python3.10/site-packages/pandas/core/tools/datetimes.py:538, in _to_datetime_with_format(arg, orig_arg, name, tz, fmt, exact, errors, infer_datetime_format)
    535         return _box_as_indexlike(result, utc=utc, name=name)
    537 # fallback
--> 538 res = _array_strptime_with_fallback(
    539     arg, name, tz, fmt, exact, errors, infer_datetime_format
    540 )
    541 return res

File ~/tmp/.venv/lib/python3.10/site-packages/pandas/core/tools/datetimes.py:473, in _array_strptime_with_fallback(arg, name, tz, fmt, exact, errors, infer_datetime_format)
    470 utc = tz == "utc"
    472 try:
--> 473     result, timezones = array_strptime(arg, fmt, exact=exact, errors=errors)
    474 except OutOfBoundsDatetime:
    475     if errors == "raise":

File ~/tmp/.venv/lib/python3.10/site-packages/pandas/_libs/tslibs/strptime.pyx:156, in pandas._libs.tslibs.strptime.array_strptime()

ValueError: unconverted data remains:  00:00:00
```

aaossa commented:
Closing in favor of #49893
